### ISSUE TITLE
[Bug]: Puppeteer fails to find Chromium on Render.com

### ISSUE THREAD

**Issue by ProchaLu** [puppeteer/puppeteer#9694]
### Bug expectation

When configuring Puppeteer using the official installation guide available at https://pptr.dev/#installation, the package should work without any additional configuration on common environments like Render.com. However, when deploying my Puppeteer application to Render.com, I encountered issues that prevented it from running correctly. 

If Puppeteer requires additional configuration for common cloud providers like Render.com and Heroku, working configurations should be included in the installation guide, or working configs for cloud providers should be included in the installation guide.

### Actual Result

Puppeteer fails to find Chromium on Render.com

- The error message suggests that the cache path for Puppeteer may be incorrectly configured. However, the configuration guide provided by Puppeteer is not specific to Render.com

- The error message provides a link to the Puppeteer configuration guide, but it is unclear what changes need to be made to the configuration for Puppeteer to work on Render.com

Issue: 
https://github.com/puppeteer/puppeteer/issues/9581 
https://github.com/puppeteer/puppeteer/issues/9592

### Bug behavior

- [ ] Flaky
- [ ] PDF

### Minimal, reproducible example

```TypeScript
import { launch } from 'puppeteer';

(async () => {
  const browser = await launch();
  const page = await browser.newPage();

  await browser.close();
})();
```

### Error string

{"message":"Could not find Chromium (rev. 1095492). This can occur if either\n 1. you did not perform an installation before running the script (e.g. npm install) or\n 2. your cache path is incorrectly configured (which is: /opt/render/.cache/puppeteer).\nFor (2), check out our guide on configuring puppeteer at [https://pptr.dev/guides/configuration."}](https://pptr.dev/guides/configuration.%22%7D)

### Puppeteer version

19.7.1

### Node version

18.13.0

### Package manager

yarn

### Package manager version

1.22.19

### Operating system

macOS

**Comment by OrKoN** [comment]
Thanks for the report. The documentation can be found here https://pptr.dev/guides/configuration#changing-the-default-cache-directory it is applicable to the variety of environments, so we won't maintain specific documentation per service.

**Comment by karlhorky** [comment]
I think this is a serious footgun in the latest versions of Puppeteer:

1. Follow the installation guide to install on your local machine -> works ✅
2. Deploy to production on Heroku, Render.com or other services -> breaks ❌ (with confusing error)

Confusing error:

```
Could not find Chromium (rev. 1095492). This can occur if either
 1. you did not perform an installation before running the script (e.g. npm install) or
 2. your cache path is incorrectly configured (which is: /opt/render/.cache/puppeteer).
For (2), check out our guide on configuring puppeteer at https://pptr.dev/guides/configuration.
```

This is confusing because:

- users will not necessarily know / remember that `"puppeteer"` in their `package.json` is causing download of Chromium
- users will be looking in their CI / hosting scripts for a missing `npm install`, `yarn install`, etc, which is actually **not** missing
- users will not know what the correct path to a "working cache path" will be
- users will not be able to find a "working cache path" anywhere in the docs, and so they will have to resort to searching through issues, and maybe finding [the buried post of Victor](https://github.com/puppeteer/puppeteer/issues/9581#issuecomment-1416057903)

This is a poor DX to make users go through.

There have already been multiple issues created about this, and there will be more coming if this isn't resolved properly.

---

Solution 1 (probably best for everyone):

Starting with a premise of stability of the ecosystem:

Puppeteer and other packages should strive to be:

- zero-configuration
- compatible with popular hosting and CI infrastructure

So the best option here would be for Puppeteer to be compatible out of the box with the way that both CI and hosting infrastructure works - no config needed.

---

Solution 2 (not great experience for developers, but avoids issue spam here):

If there is a configuration option that should be applied to CI and hosting environments, then call this out in the docs within [the `Installation` section](https://pptr.dev/#installation) (not the `Configuration` subheading, which implies it's a choice) to show that it's a mandatory step to not have Puppeteer crash and burn when deployed to common hosting providers.

**Comment by OrKoN** [comment]
@karlhorky what is your idea about where the browser should be downloaded to by default? Previously, we had it in the node_modules but had another stream of bugs where after the code was bundled and moved to the another machine, the browser would not be found as well. Installing in the current working directory also sounds wasteful and, there is a risk that binaries will end up being committed to the repo and cause problems due to the platform mismatch. Actually, do you know why does it not work in render.com and similar services? it works on GitHub actions, for example. Would it make sense to open an issue with those vendors so that they include .cache folders into the final builds? The .cache folder is a common path where the modules are installed. 

We are also working on a CLI to manage the browsers better. We could implement the solution 2 but I am not sure it will stop the issue from being reported.

**Comment by karlhorky** [comment]
Thanks for the quick answer! Hearing the details about the challenges helps to understand why other options work less well 👍 

I'm not sure what I would suggest for a location - I'm not sure I have a full understanding of the constraints to observe here yet. Wonder if there could be a check for commonly-set environment variables (eg. `CI=true` or `NODE_ENV=production`) during installation and the current app directory used as an installation instead? ([Render sets these](https://render.com/docs/environment-variables))

> Actually, do you know why does it not work in render.com and similar services?

I'll chat with @ProchaLu - we'll try to find out more details about what Render.com is doing and post back here. Seems like Heroku is also having this problem.

> Would it make sense to open an issue with those vendors so that they include .cache folders into the final builds?

If I'm understanding [the `v19.0.0` release notes](https://github.com/puppeteer/puppeteer/releases/tag/v19.0.0) correctly, the path where the browsers are being installed is `~/.cache/puppeteer` - maybe these hosting providers have a problem with the home directory (vs the directory where the app is installed).

cc @anurag, @aduong or @ezquire from Render in case you would have any input here

**Comment by OrKoN** [comment]
It's a bit hard to strike a balance here: for example, CI=true is also set by GitHub Actions but the cache folder is resolved without problems there, and it allows for [efficient caching of binaries between workflows](https://github.com/puppeteer/puppeteer/blob/main/.github/workflows/ci.yml#L218). I believe what might be happening at Render.com is that either the home directory is different between the build time and the runtime or the files from home directly are included in the final image so the cache folder is empty at runtime.

I have opened a PR to explicitly call out this nuance: https://github.com/puppeteer/puppeteer/pull/9695

**Comment by ProchaLu** [comment]
Thanks for your comments and quick answers, @OrKoN. I wanted to let you know that I've made progress on my issue with Puppeteer and Render.com.

Following your suggestion and a comment from Victor on [this Puppeteer issue](https://github.com/puppeteer/puppeteer/issues/9581#issuecomment-1416057903), I created a `puppeteer.config.cjs` file with the following code:

```js
const { join } = require('path');

/**
 * @type {import("puppeteer").Configuration}
 */
module.exports = {
  // Changes the cache location for Puppeteer.
  cacheDirectory: join(__dirname, '.cache', 'puppeteer'),
};
```
I then uninstalled and reinstalled Puppeteer and deployed my project on Render.com. I'm happy to report that everything now works as it should.

Thanks again for your help.

If anyone else has the same issue when deploying on Render.com, here is a small [example repository](https://github.com/ProchaLu/render-puppeteer-test)

**Comment by ProchaLu** [comment]
Hi @OrKoN,

I am currently working on a monorepo example and using `puppeteer.config.cjs` in the root, I have successfully created a `.cache` folder, and everything is working as expected. However, I now want to switch to a Typescript config file called puppeteer.config.ts.

In the puppeteer.config.ts file, I have specified the cacheDirectory using the join method from the path module and exported the configuration as `puppeteerConfig`. Here's the code snippet:
```
import { join } from 'path';
import { Configuration } from 'puppeteer';

const puppeteerConfig: Configuration = {
  cacheDirectory: join(__dirname, '.cache', 'puppeteer'),
};

export default puppeteerConfig;
```
My question is, how can I configure npm install/yarn install to respect the cacheDirectory config in `puppeteer.config.ts`? Currently, the deployment works for the puppeteer-test folder using the `puppeteer.config.cjs` file, but the `.cache` folder is not being created with the Typescript config.

You can check out my example repo [here](https://github.com/ProchaLu/render-puppeteer-monorepo-test). Thank you for your help!

**Comment by ProchaLu** [comment]
Although the documentation doesn't provide an example for a Typescript config file, there might still be a way to implement the configuration using a .ts file, that's why I am asking
Thanks!

**Comment by ProchaLu** [comment]
Render.com deployment fails because since Puppeteer `v19.0.0`, Puppeteer downloads and installs Chromium by default in `$HOME/.cache/puppeteer`, which is not present in the project's production environment. Without adding a config file to the root of the project, the deployment on Render.com will throw an error when running Puppeteer:

```
"Could not find Chromium (rev. 1095492). This can occur if either
1. you did not perform an installation before running the script (e.g. npm install) or
2. your cache path is incorrectly configured (which is: /opt/render/.cache/puppeteer)."
```

To further investigate the issue, I created [this reproduction repo](https://github.com/ProchaLu/render-puppeteer-test) and ran scripts in the build environment to see the different paths, logging the executable path for Puppeteer and the path to Chrome with the command:

```bash
$node -e "console.log(require('puppeteer').executablePath())"
/opt/render/.cache/puppeteer/chrome/linux-1095492/chrome-linux/chrome

$find / -name chrome | grep puppeteer | grep -v node_modules
/opt/render/.cache/puppeteer/chrome
/opt/render/.cache/puppeteer/chrome/linux-1095492/chrome-linux/chrome
```

However, running the same commands in the Render.com production environment resulted in an error for the executable path and no path for Chrome.

```
Feb 27 10:30:03 AM /opt/render/project/src/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:127
Feb 27 10:30:03 AM throw new Error(`Could not find Chromium (rev. ${this.puppeteer.browserRevision}). This can occur if either\n` +
Feb 27 10:30:03 AM ^
Feb 27 10:30:03 AM
Feb 27 10:30:03 AM Error: Could not find Chromium (rev. 1095492). This can occur if either
Feb 27 10:30:03 AM 1. you did not perform an installation before running the script (e.g. `npm install`) or
Feb 27 10:30:03 AM 2. your cache path is incorrectly configured (which is: /opt/render/.cache/puppeteer).
Feb 27 10:30:03 AM For (2), check out our guide on configuring puppeteer at https://pptr.dev/guides/configuration.
Feb 27 10:30:03 AM at ChromeLauncher.resolveExecutablePath (/opt/render/project/src/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:127:27)
Feb 27 10:30:03 AM at ChromeLauncher.executablePath (/opt/render/project/src/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:206:25)
Feb 27 10:30:03 AM at PuppeteerNode.executablePath (/opt/render/project/src/node_modules/puppeteer-core/lib/cjs/puppeteer/node/PuppeteerNode.js:162:105)
```

I looked through the Render.com documentation on how to copy over specific files from the build environment to the production environment but did not find anything. The section [Deploy Puppeteer with Node](https://render.com/docs/deploy-puppeteer-node) also does not offer any specific configuration for Puppeteer.

I reported the problem of Puppeteer failing to find Chromium on Render.com in [this issue](https://github.com/puppeteer/puppeteer/issues/9694), which led to @OrKon [updating the Puppeteer docs](https://github.com/puppeteer/puppeteer/pull/9695) with information related to the problem.

While seeking a solution to the deployment issue, I came across a related problem encountered by @VictorAssis [deploying Puppeteer to Heroku](https://github.com/puppeteer/puppeteer/issues/9581#issuecomment-1416057903) on the Puppeteer repository. To address the error, I created a `puppeteer.config.cjs` file in the project's root directory, configuring the `cacheDirectory` option to specify the location for Puppeteer's cached data. You can reference an example of this configuration file in the [documentation](https://pptr.dev/guides/configuration#examples).

```js
const { join } = require('path');

/**
 * @type {import("puppeteer").Configuration}
 */
module.exports = {
  cacheDirectory: join(__dirname, '.cache', 'puppeteer'),
};
```

By setting the `cacheDirectory` option to `__dirname/.cache/puppeteer` in the `puppeteer.config.cjs` file at the root of the project, it is specifying a cache directory that Puppeteer should use for storing and accessing Chromium. By using the same cache directory for both environments, we can ensure that Puppeteer can find the correct path for Chromium when deploying on Render.com. This resolved the issue of Puppeteer being unable to find the Chromium browser in the production environment. To confirm that Chromium was installed at the same path, I ran the same commands and verified that it was located at `/opt/render/project/src/.cache/puppeteer/chrome/linux-1095492/chrome-linux/chrome`.

```bash
$node -e "console.log(require('puppeteer').executablePath())"
/opt/render/project/src/.cache/puppeteer/chrome/linux-1095492/chrome-linux/chrome

$find / -name chrome | grep puppeteer | grep -v node_modules
/opt/render/project/src/.cache/puppeteer/chrome
/opt/render/project/src/.cache/puppeteer/chrome/linux-1095492/chrome-linux/chrome
```

cc @aduong, @AnalogMemory, @anurag, @jimthoburn, @ohbh

**Comment by karlhorky** [comment]
I've also asked @jadepaoletta whether Render would consider syncing / copying `$HOME/.cache` to the production environment over here:

- https://community.render.com/t/docker-puppeteer-not-working-on-a-express-nestjs-app/2807/7

**Comment by BigAlUK** [comment]
Hi there, I'm a member of the Render support team. We do get a number of tickets and forum posts on this subject, but I don't think it's the Puppeteer maintainers responsibility to worry about where their package is deployed.

TL;DR; On Render, install Chrome at build time, to a path that is carried over from the build to runtime.

The quickest solution I found for this was setting the `PUPPETEER_CACHE_DIR` environment variable  (_[docs](https://pptr.dev/api/puppeteer.configuration.cachedirectory/)_) to `/opt/render/project/.puppeteer`

However, as the `node_packages` directory is cached by default (and the path above isn't), the Puppeteer-included Chrome installer doesn't run if the next deploy uses the already installed `node_modules` in the build cache. The immediate workaround would be to do a manual deploy with a clear build cache, so that the Chrome installer runs when the Puppeteer package is reinstalled. But that's not a great user experience.

It's on my list to add a Puppeteer guide to the Render docs and/or an example repo to https://github.com/render-examples. It wouldn't be much more than what I've described above, but I plan to include a simple build script to check Chrome is installed and also added to/pulled from the Render service build cache. I don't have an ETA, it depends on my other workload.

It would be unlikely that we would add any auto copying/syncing features, we keep our Native Environments predictable and flexible enough to achieve a lot of customization with a build script, there's no "magic" to reduce anything unexpected.

**Comment by karlhorky** [comment]
> It would be unlikely that we would add any auto copying/syncing features, we keep our Native Environments predictable and flexible enough to achieve a lot of customization with a build script, there's no "magic" to reduce anything unexpected.

Oh, too bad. Seems like (according to @OrKoN) `$HOME/.cache` is a somewhat standard location for a cache, with GitHub Actions and others persisting this location across build and production environments (also is the same as the `$XDG_CACHE_HOME` environment variable, which seems to be a standard that Render at least partially observes).

So I guess it's Render's opinion that "if some process installs some files during build outside of the app directory `/opt/render/project/src`, these files will be discarded in the production environment".

**Comment by nani1618** [comment]
@karlhorky Any Solution to the issue?? I am trying to deploy my node.js application on Render.com and getting the same error. Help me if your problem is solved

**Comment by zahinz** [comment]
Thank you so much @ProchaLu ! 

> Hi @OrKoN,
> 
> I am currently working on a monorepo example and using `puppeteer.config.cjs` in the root, I have successfully created a `.cache` folder, and everything is working as expected. However, I now want to switch to a Typescript config file called puppeteer.config.ts.
> 
> In the puppeteer.config.ts file, I have specified the cacheDirectory using the join method from the path module and exported the configuration as `puppeteerConfig`. Here's the code snippet:
> 
> ```
> import { join } from 'path';
> import { Configuration } from 'puppeteer';
> 
> const puppeteerConfig: Configuration = {
>   cacheDirectory: join(__dirname, '.cache', 'puppeteer'),
> };
> 
> export default puppeteerConfig;
> ```
> 
> My question is, how can I configure npm install/yarn install to respect the cacheDirectory config in `puppeteer.config.ts`? Currently, the deployment works for the puppeteer-test folder using the `puppeteer.config.cjs` file, but the `.cache` folder is not being created with the Typescript config.
> 
> You can check out my example repo [here](https://github.com/ProchaLu/render-puppeteer-monorepo-test). Thank you for your help!

**Comment by ewarrenG** [comment]
hi I was recieving this same error, so far I have added done the following:

1. added`puppeteer.config.cjs` file and uninstalled and reinstalled puppeteer. I also added the `.cache` to my ``.gitignore` -- adding this since it wasn't mentioned before.
2. added an environment variable inside my render web service for `PUPPETEER_CACHE_DIR` and set it to `/opt/render/project/.puppeteer` and triggered a manual deploy to clear the cache as suggested by @BigAlUK. 

however, it's still not working, in the logs of my application I am seeing

```
  e: Error: Failed to launch the browser process! undefined
  [76:91:1222/161141.574655:ERROR:bus.cc(407)] Failed to connect to the bus: Failed to connect to socket /var/run/dbus/system_bus_socket: No such file or directory
  [76:76:1222/161141.699558:ERROR:ozone_platform_x11.cc(239)] Missing X server or $DISPLAY
  [76:76:1222/161141.699598:ERROR:env.cc(255)] The platform failed to initialize.  Exiting.
  ```

I am not using docker but have specified my environment variables as a secret file. Can someone assist me further?

**Comment by tkwn2080** [comment]
Hi, I've had success with placing the config file in the root. However whenever I push an update to auto-deploy through GitHub, then it throws an error again on loading—but then this is cured by manually deploying with a cleared build cache. While this is certainly better than not working, it is also far from perfect. I was wondering whether anyone (especially at Render, e.g., @BigAlUK) might have an idea for how to avoid this manual step requirement. Thanks.

**Comment by hafaiedhmehdy** [comment]
None of the above options work for me; it's so weird. 😒

```                    throw new Error(`Could not find Chrome (ver. ${this.puppeteer.browserRevision}). This can occur if either\n` +
                          ^
Error: Could not find Chrome (ver. 119.0.6045.105). This can occur if either```

Could anyone post a sample of a basic repository that actually works? I'd like to fork it and give it a try to see what's different from my use case. Thanks! 😍

**Comment by whackatracker** [comment]
> node_modules/puppeteer/install.js

Thanks, I tried this but it didn't work since my `node_modules` are excluded from my github repo. Will keep trying the other solutions here...

**Comment by marcelEzenagu** [comment]
https://github.com/puppeteer/puppeteer/issues/9694#issuecomment-1448664518

Did the work for "puppeteer": "^19.11.1"
